File: src/flet_pro_image_editor/flet_pro_image_editor.py
import flet as ft
import flet.version
from flet.core.control import Control, OptionalNumber
from flet.core.ref import Ref
from typing import Any, Callable, Optional, Union, List, Dict


class ProImageEditor(ft.Control):
    def __init__(
        self,
        show_status: bool = False,
        on_editing_complete=None,
        on_editing_cancel=None,
        #
        ref: Optional[Ref] = None,
        disabled: Optional[bool] = None,
        visible: Optional[bool] = None,
        data: Any = None,
    ):
        super().__init__(ref=ref, disabled=disabled, visible=visible, data=data)
        self.show_status = show_status
        self._source = None
        self._path = None
        self._crop_ratio = None
        
        # Registrar los manejadores de eventos directamente
        self._add_event_handler("editing_complete", on_editing_complete)
        self._add_event_handler("editing_cancel", on_editing_cancel)

    def _get_control_name(self):
        return "flet_pro_image_editor"

    def _get_children(self):
        return []

    def _before_build_command(self):
        super()._before_build_command()

    def _build(self):
        args = {}
        if self.show_status:
            args["showStatus"] = "true"
        return super()._build() | args

    def open_editor(self, source: str, path: str, crop_ratio: str = None):
        """
        Abre el editor de imágenes con la imagen especificada.

        Args:
            source: Fuente de la imagen ('network', 'file', 'asset')
            path: Ruta a la imagen
            crop_ratio: Relación de aspecto para el recorte (ej. '1.0', '4/3', '16/9')
        """
        self._source = source
        self._path = path
        self._crop_ratio = crop_ratio

        args = {"source": source, "path": path}
        if crop_ratio:
            args["crop_ratio"] = crop_ratio

        return self.invoke_method("open_editor", args)

    def close_editor(self):
        """
        Cierra el editor de imágenes.
        """
        return self.invoke_method("close_editor", {})

    def open(self, dialog):
        """
        Abre el editor de imágenes como un diálogo.

        Args:
            dialog: El control ProImageEditor a abrir
        """
        if not self._source or not self._path:
            raise ValueError(
                "Debe llamar a open_editor primero para configurar la imagen"
            )

        args = {"source": self._source, "path": self._path}
        if self._crop_ratio:
            args["crop_ratio"] = self._crop_ratio

        return self.invoke_method("open", args)

    def close(self, dialog):
        """
        Cierra el editor de imágenes.

        Args:
            dialog: El control ProImageEditor a cerrar
        """
        return self.invoke_method("close", {})
        
    # Implementación simplificada de los manejadores de eventos
    @property
    def on_editing_complete(self):
        return self._get_event_handler("editing_complete")
        
    @on_editing_complete.setter
    def on_editing_complete(self, handler):
        self._add_event_handler("editing_complete", handler)
        
    @property
    def on_editing_cancel(self):
        return self._get_event_handler("editing_cancel")
        
    @on_editing_cancel.setter
    def on_editing_cancel(self, handler):
        self._add_event_handler("editing_cancel", handler)

File: src/flutter/flet_pro_image_editor/lib/src/flet_pro_image_editor.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:pro_image_editor/pro_image_editor.dart';
import 'dart:typed_data';

class FletProImageEditorControl extends StatefulWidget {
  final Control? parent;
  final Control control;
  final List<Control> children;
  final bool parentDisabled;
  final bool? parentAdaptive;
  final FletControlBackend backend;

  const FletProImageEditorControl({
    super.key,
    required this.parent,
    required this.control,
    required this.children,
    required this.parentDisabled,
    required this.parentAdaptive,
    required this.backend,
  });

  @override
  State<FletProImageEditorControl> createState() =>
      _FletProImageEditorControlState();
}

class _FletProImageEditorControlState extends State<FletProImageEditorControl> {
  bool _isEditing = false;
  Uint8List? _editedImage;
  bool _isOpen = false;

  @override
  void initState() {
    super.initState();
    widget.backend.subscribeMethods(widget.control.id, _handleMethodCall);
  }

  @override
  void dispose() {
    widget.backend.unsubscribeMethods(widget.control.id);
    super.dispose();
  }

  Future<String?> _handleMethodCall(
      String methodName, Map<String, String> args) async {
    switch (methodName) {
      case "open_editor":
        return _openEditor(args);
      case "close_editor":
        return _closeEditor();
      case "open":
        return _openDialog(args);
      case "close":
        return _closeDialog();
      default:
        return null;
    }
  }

  Future<String?> _openDialog(Map<String, String> args) async {
    // Cerrar cualquier diálogo existente primero
    if (_isOpen && context.mounted) {
      Navigator.of(context, rootNavigator: true).pop();
    }

    setState(() {
      _isOpen = true;
    });

    // Obtener configuraciones del editor
    String? imageSource = args["source"];
    String? imagePath = args["path"];

    if (imageSource == null || imagePath == null) {
      return "error: Se requiere source y path";
    }

    // Configuraciones del editor
    ProImageEditorConfigs configs = ProImageEditorConfigs(
      cropRotateEditor: args["crop_ratio"] != null
          ? CropRotateEditorConfigs(
              initAspectRatio: double.tryParse(args["crop_ratio"]!),
            )
          : const CropRotateEditorConfigs(),
      mainEditor: const MainEditorConfigs(),
    );

    // Crear el widget del editor según la fuente
    Widget editorWidget;
    if (imageSource == "network") {
      editorWidget = ProImageEditor.network(
        imagePath,
        configs: configs,
        callbacks: ProImageEditorCallbacks(
          onImageEditingComplete: _handleEditingComplete,
          onCloseEditor: (_) async => await _handleEditingCancel(),
        ),
      );
    } else if (imageSource == "file") {
      editorWidget = ProImageEditor.file(
        imagePath,
        configs: configs,
        callbacks: ProImageEditorCallbacks(
          onImageEditingComplete: _handleEditingComplete,
          onCloseEditor: (_) async => await _handleEditingCancel(),
        ),
      );
    } else if (imageSource == "asset") {
      editorWidget = ProImageEditor.asset(
        imagePath,
        configs: configs,
        callbacks: ProImageEditorCallbacks(
          onImageEditingComplete: _handleEditingComplete,
          onCloseEditor: (_) async => await _handleEditingCancel(),
        ),
      );
    } else {
      return "error: Fuente de imagen no válida. Use 'network', 'file' o 'asset'";
    }

    // Mostrar el editor como un diálogo
    if (context.mounted) {
      showDialog(
        context: context,
        barrierDismissible: false,
        builder: (BuildContext context) {
          return PopScope(
            canPop: false,
            onPopInvoked: (didPop) async {
              if (!didPop) {
                await _handleEditingCancel();
              }
            },
            child: Dialog(
              insetPadding: EdgeInsets.zero,
              child: Container(
                width: double.infinity,
                height: double.infinity,
                color: Colors.black,
                child: editorWidget,
              ),
            ),
          );
        },
      );
    }

    return "success";
  }

  Future<String?> _closeDialog() async {
    if (_isOpen && context.mounted) {
      Navigator.of(context, rootNavigator: true).pop();
      setState(() {
        _isOpen = false;
      });
    }
    return "success";
  }

  Future<String?> _openEditor(Map<String, String> args) async {
    // Asegurarse de que el estado se reinicie correctamente antes de abrir el editor
    // Limpiar explícitamente la imagen editada y establecer el estado de edición
    setState(() {
      _isEditing = true;
      _editedImage = null;
    });

    // Asegurar que cualquier cambio de estado se aplique antes de continuar
    await Future.delayed(Duration.zero);

    try {
      // Determinar la fuente de la imagen
      String? imageSource = args["source"];
      String? imagePath = args["path"];

      if (imageSource == null || imagePath == null) {
        return "error: Se requiere source y path";
      }

      // Configuraciones del editor
      ProImageEditorConfigs configs = ProImageEditorConfigs(
        // Configuración de CropRotateEditor para manejar la relación de aspecto
        cropRotateEditor: args["crop_ratio"] != null
            ? CropRotateEditorConfigs(
                initAspectRatio: double.tryParse(args["crop_ratio"]!),
              )
            : const CropRotateEditorConfigs(),
        // Configuración de MainEditor
        mainEditor: const MainEditorConfigs(),
      );

      // Nota: Las propiedades availableOperations e initialOperation ya no están disponibles en la API actual

      // Abrir el editor según la fuente
      Widget editorWidget;
      if (imageSource == "network") {
        editorWidget = ProImageEditor.network(
          imagePath,
          configs: configs,
          callbacks: ProImageEditorCallbacks(
            onImageEditingComplete: _handleEditingComplete,
            onCloseEditor: (_) async => await _handleEditingCancel(),
          ),
        );
      } else if (imageSource == "file") {
        editorWidget = ProImageEditor.file(
          imagePath,
          configs: configs,
          callbacks: ProImageEditorCallbacks(
            onImageEditingComplete: _handleEditingComplete,
            onCloseEditor: (_) async => await _handleEditingCancel(),
          ),
        );
      } else if (imageSource == "asset") {
        editorWidget = ProImageEditor.asset(
          imagePath,
          configs: configs,
          callbacks: ProImageEditorCallbacks(
            onImageEditingComplete: _handleEditingComplete,
            onCloseEditor: (_) async => await _handleEditingCancel(),
          ),
        );
      } else {
        return "error: Fuente de imagen no válida. Use 'network', 'file' o 'asset'";
      }

      // Mostrar el editor en una nueva página
      await Navigator.push(
        context,
        MaterialPageRoute(builder: (context) => editorWidget),
      );

      return "success";
    } catch (e) {
      debugPrint("Error al abrir el editor: $e");
      setState(() {
        _isEditing = false;
      });
      return "error: $e";
    }
  }

  // Asegurarse de que los eventos se disparen con el nombre correcto
  Future<void> _handleEditingComplete(Uint8List editedImage) async {
    // Primero disparamos el evento antes de cerrar el diálogo
    List<int> bytesList = editedImage.toList();
    String bytesString = bytesList.join(',');

    // Disparar el evento con el nombre correcto
    widget.backend.triggerControlEvent(
        widget.control.id, "editing_complete", bytesString);

    // Luego actualizamos el estado
    setState(() {
      _editedImage = editedImage;
      _isEditing = false;
    });

    // Cerrar el diálogo si está abierto
    if (_isOpen && context.mounted) {
      Navigator.of(context, rootNavigator: true).pop();
      setState(() {
        _isOpen = false;
      });
    }
  }

  Future<void> _handleEditingCancel() async {
    // Primero disparamos el evento antes de cerrar el diálogo
    widget.backend.triggerControlEvent(widget.control.id, "editing_cancel", "");

    // Luego actualizamos el estado
    setState(() {
      _isEditing = false;
    });

    // Cerrar el diálogo si está abierto
    if (_isOpen && context.mounted) {
      Navigator.of(context, rootNavigator: true).pop();
      setState(() {
        _isOpen = false;
      });
    }

    // Asegurarse de cerrar la página del editor si está abierta
    if (Navigator.of(context).canPop() && context.mounted) {
      Navigator.of(context).pop();
    }
  }

  Future<String?> _closeEditor() async {
    // Cerrar la página del editor si está abierta
    if (context.mounted && Navigator.of(context).canPop()) {
      Navigator.of(context).pop();
    }

    // Cerrar el diálogo si está abierto
    if (_isOpen && context.mounted) {
      Navigator.of(context, rootNavigator: true).pop();
      setState(() {
        _isOpen = false;
      });
    }

    // Actualizar el estado local y limpiar la imagen editada
    setState(() {
      _isEditing = false;
      _editedImage = null; // Asegurarse de limpiar la imagen editada
    });

    // Asegurar que los cambios de estado se apliquen completamente
    await Future.delayed(Duration.zero);

    return "success";
  }

  // Nota: Los métodos _parseAvailableOperations y _parseInitialOperation han sido eliminados
  // ya que las propiedades availableOperations e initialOperation no están disponibles en la API actual

  @override
  Widget build(BuildContext context) {
    // Este control no tiene representación visual directa

    return const SizedBox.shrink(); // Control invisible
  }
}
